craft_application/commands/lifecycle.py

```diff
@@ -416,11 +429,88 @@ def _run_real(
             emit.progress(f"Packed: {package_names}", permanent=True)
             artifact, resources = packages[0], self._services.package.resource_map
 
+        self._save_packed_file_list(artifact=artifact, resources=resources)
         self._services.package.write_state(artifact=artifact, resources=resources)
 
         if shell_after:
             _launch_shell()
 
+    def _is_already_packed(self) -> bool:
+        """Verify whether the artifacts are already packed and up-to-date."""
+        # 1. A file containing the list of packed artifacts is created after
+        #    packing. Before packing, check if the list file exists. If not, we
+        #    never packed before and packing is necessary.
+        # 2. Check the most recent prime state timestamp. These are generated
+        #    when parts are primed. If this information can't be retrieved, we
+        #    repack.
+        # 3. If any part was primed after the artifacts were generated, we need
+        #    to pack again. Note that if there's any change in parts, the part
+        #    will be reprimed before the pack step and artifacts will be repacked.
+        # 4. If any previously packed file is missing, we repack everything.
+        # 5. Otherwise, the existing files are up-to-date and no repacking is
+        #    required.
+
+        pack_time = self._get_packed_file_list_timestamp()
+        if pack_time is None:
+            return False
+
+        prime_time = self._services.lifecycle.prime_state_timestamp
+        if prime_time is None:
+            # This should never happen under normal circumstances, but manual
+            # manipulation of state files is possible.
+            return False
```

Comment:
Since this isn't an error but also never should happen, we should help our future selves with a log message:

```suggestion
            # This should never happen under normal circumstances, but manual
            # manipulation of state files is possible.
            emit.debug("Couldn't find prime state timestamps.")
            return False
```